fix: serialize decimals as quoted strings in JSON output

okhsunrog opened the pull request:
For correct serialization of decimal types into JSON we want them to be serialized as strings instead of numbers so they don't loose precision.

novartole commented on arrow-json/src/writer/encoder.rs:
```diff
@@ -141,7 +141,7 @@ fn make_encoder_impl<'a>(
             let options = FormatOptions::new().with_display_error(true);
             let formatter = ArrayFormatter::try_new(array, &options)?;
-            (Box::new(RawArrayFormatter(formatter)) as _, array.nulls().cloned())
+            (Box::new(formatter) as _, array.nulls().cloned())
```
What about to extend `FormatOptions` instead? This allows users to choose the format and keeps backward compatibility.

okhsunrog replied:
Done, could you take a look?

askalt commented:
Thank you for the patch!

We need to fix tests here:
https://github.com/tarantool/arrow-rs/blob/main/arrow-json/src/writer/mod.rs

Btw, does decimal **de**serializer properly works with input strings?

okhsunrog commented:
I use `parse_decimal` from `arrow-cast/src/parse.rs` and it works good

askalt commented:
> I use `parse_decimal` from `arrow-cast/src/parse.rs` and it works good

Ok, it seems that deserializer also uses it https://github.com/tarantool/arrow-rs/blob/main/arrow-json/src/reader/decimal_array.rs#L57-L60, so it will be able to deserialize strings as decimals.

novartole commented on arrow-cast/src/display.rs:
```diff
@@ -72,6 +72,9 @@ pub struct FormatOptions<'a> {
     /// Duration format
     duration_format: DurationFormat,
+    /// If set to `true`, decimal values will be formatted as strings instead of numbers
+    /// This is useful for JSON output to preserve precision
```
It could be ignored, but I see that two close sentences usually end up with dots.
```suggestion
    /// If set to `true`, decimal values will be formatted as strings instead of numbers.
    /// This is useful for JSON output to preserve precision.
```

novartole commented on arrow-cast/src/display.rs:
```diff
@@ -72,6 +72,9 @@ pub struct FormatOptions<'a> {
     duration_format: DurationFormat,
+    /// If set to `true`, decimal values will be formatted as strings instead of numbers
+    /// This is useful for JSON output to preserve precision
+    decimal_as_string: bool,
```
Does it make sense to change it to enum of formats instead as it's done for others?

okhsunrog replied:
Good idea, I refactored it to use an enum.

novartole commented on arrow-cast/src/display.rs:
```diff
@@ -38,6 +38,22 @@ use lexical_core::FormattedSize;
+    fn default() -> Self {
+        Self::Number
+    }
+}
```
A bit shorter, but still explicit:
```suggestion
/// Format for displaying decimals
#[derive(Default, Debug, Copy, Clone, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum DecimalFormat {
    /// Render decimals as JSON numbers, e.g. 12.34
    #[default]
    Number,
    /// Render decimals as JSON strings, e.g. "12.34"
    String,
}
```

novartole commented on arrow-json/src/writer/encoder.rs:
```diff
@@ -19,16 +19,26 @@ use arrow_array::cast::AsArray;
+            explicit_nulls: false,
+            decimal_format: DecimalFormat::Number,
+        }
+    }
```
```suggestion
#[derive(Debug, Clone, Default)]
pub struct EncoderOptions {
    pub explicit_nulls: bool,
    pub decimal_format: DecimalFormat,
}
```

novartole commented on arrow-json/src/writer/mod.rs:
```diff
@@ -1938,18 +1968,116 @@ mod tests {
-            let mut writer = LineDelimitedWriter::new(&mut buf);
-            writer.write_batches(&[&batch]).unwrap();
+    #[test]
+    fn test_decimal_list_encoder_as_string() {
```
It seems `*_as_string` doesn't make any sense, because the both outputs are asserted anyway. And let's rename the test to reflect the target type.
```suggestion
    fn test_decimal128_list_encoder() {
```
If you agree, let's apply this to the rest:
- `test_decimal_dictionary_encoder_as_string` -> `test_decimal128_dictionary_encoder`
- `test_decimal256_dictionary_encoder_as_string` -> `test_decimal256_dictionary_encoder`

okhsunrog replied:
Done

novartole commented:
Thank you for the patch!

The last, but not least: let's squash commits keeping only meaningful ones (perhaps, reduce to one, but it's up to you) and add a bit of context about things are done.

The rest LGTM.

okhsunrog committed:
feat: add configurable decimal serialization format for JSON writer

Add DecimalFormat enum with Number and String variants to control how
decimal values are serialized in JSON output. The default Number format
renders decimals as JSON numbers (e.g., 12.34), while String format
renders them as quoted strings (e.g., "12.34").

Changes:
- Add DecimalFormat enum to arrow-cast with Number (default) and String variants
- Add decimal_format field to EncoderOptions and FormatOptions
- Implement decimal formatting in DisplayIndexState for Decimal128/256
- Add WriterBuilder::with_decimal_format() configuration method
- Add tests for decimal arrays, lists, and dictionaries
- Fix lifetime elision warnings in BitChunks methods

okhsunrog commented:
@novartole thank you for the review! I squashed the commits and created a more meaningful commit message.

askalt reviewed (commented):
Thank you! LGTM
